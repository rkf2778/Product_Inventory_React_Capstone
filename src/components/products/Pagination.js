import React from 'react'
import { Link } from "react-router-dom";


export default function Pagination({postsPerPage,totalPosts,paginate}) {
    const pageNumbers = []

    for(let i=1;i<=Math.ceil(totalPosts/postsPerPage);i++){
        pageNumbers.push(i)
    }
    return (
        <nav>
           <ul className="pagination">
                {pageNumbers.map(number=>(
                    <li key={number} className="page-item">
                        <Link className="page-link" onClick={()=>paginate(number)} >
                            {number}
                        </Link>
                    </li>
                ))}
           </ul>    
        </nav>
    )
}
