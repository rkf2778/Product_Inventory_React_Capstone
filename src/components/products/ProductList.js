import React from "react";
import Product from "./Product";
import { Container, Row, Col} from "react-bootstrap";

const chunk = (arr, chunkSize = 1, cache = []) => {
  const tmp = [...arr]
  if (chunkSize <= 0) return cache
  while (tmp.length) cache.push(tmp.splice(0, chunkSize))
  return cache
}

const ProductList = (props) => {
  const productsChunks = chunk(props.products, 3)
  
  const rows = productsChunks.sort( (a,b)=>(a.id>b.id)?1:-1 ).map((productChunk, index) => {
        const productsCols = productChunk.map((product, index) => {
          return (
            <Col xs="auto" sm="auto" md="auto" lg="auto" key={product.id} style={{"paddingBottom":"20px"}}>
              <Product 
              key={product.id} 
              id={product.id}
              quantity={product.quantity} 
              price={product.price} 
              name={product.name} 
              description={product.description}
              manufacturer={product.manufacturer}
              views={product.views}
                {...props}
              />	  
            </Col>
          );
        });
    return (
      <Row key={index} style={{"paddingBottom":"20px"}}>
       {productsCols}
      </Row>
            
  )});
	return (
  	<Container>
  	  {rows}
  	</Container>
  )
}

export default ProductList;
