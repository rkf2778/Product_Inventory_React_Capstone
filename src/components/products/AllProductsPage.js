import React, { useState } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Link } from "react-router-dom";
import ProductList from "./ProductList";
import Pagination from './Pagination'
import * as productActions from "../redux/actions/productActions";
import * as userActions from '../redux/actions/userActions'
import { Button } from "react-bootstrap";
import {FiSearch} from 'react-icons/fi'
import { Container, Row, Col} from "react-bootstrap";

const AllProductsPage =(props)=> {
    const [quantity, showQuantity] = useState(true);
    const [price, showPrice] = useState(true);
    const [manufacturer,showManufacturer] = useState(true);
    const data = {quantity,price,manufacturer};
    const [search,setSearch]=useState("");
    const loggedIn = props.loggedIn;

    
    //Pagination Logic
    const [currentPage,setCurrentPage] = useState(1)
    const postsPerPage = 9
    const indexOfLastPost = currentPage * postsPerPage;
    const indexOfFirstPost = indexOfLastPost - postsPerPage;
    const currentPosts = props.products.slice(indexOfFirstPost,indexOfLastPost)

    //Change the page
    const paginate =(pageNumber)=>{
      if(currentPage!==pageNumber)setCurrentPage(pageNumber)
    }


     //const filteredSearch = props.products && props.products.filter(product=>product.name.toLowerCase().indexOf(search.toLowerCase())!==-1).sort( (a,b)=>(a.id>b.id)?1:-1 );
    const filteredSearch = currentPosts.filter(product=>
                            product.name.toLowerCase().indexOf(search.toLowerCase())!==-1).sort( (a,b)=>(a.id>b.id)?1:-1 );

    return (
      <div>

        <div style={{"display":"flex","paddingTop":"30px"}} className="container">
         { loggedIn && <Link to="/addProduct"><Button variant="primary">Add Product</Button>{" "}</Link> }

          <span style={{"marginLeft":"auto"}}><input type="text" onChange={event=>setSearch(event.target.value)}/> {" "} <FiSearch size="20px"/> </span>
        </div>

        <div style={{"display":"flex","justifyContent":"flex-end","alignItems":"space-between","paddingTop":"6px"}} className="container" >
          <label style={{"padding":"0px 5px 0px 2px","color":"white"}}><input type="checkbox"  defaultChecked={quantity} onClick={()=>showQuantity(!quantity)}/>{" "}Quantity</label>
          <label style={{"padding":"0px 5px 0px 2px","color":"white"}}><input type="checkbox"  defaultChecked={price} onClick={()=>showPrice(!price)}/>{" "}Price </label>
          <label style={{"padding":"0px 5px 0px 2px","color":"white"}}><input type="checkbox"  defaultChecked={manufacturer} onClick={()=>showManufacturer(!manufacturer)}/>{" "}Manufacturer </label>
        </div>
        
        <hr></hr>

        <div style={{minHeight:"100vh"}}>
        <ProductList 
          products={filteredSearch} 
          data={data} 
          togglePrice={showPrice}
          toggleQuantity={showQuantity}
          toggleManufacturer={showManufacturer}
          loggedIn={props.loggedIn}
        />
        <br />
        <Container>
          <Row>
          <Col></Col>
          <Col xs="auto" sm="auto" md="auto" lg="auto">
            <Pagination postsPerPage={postsPerPage} totalPosts={props.products.length} paginate={paginate} />
          </Col>
          <Col></Col>
          </Row>
        </Container>
        </div>
      </div>
    );
}

function mapStateToProps(state, ownProps) {
  return {
    products: state.products,
    users : state.users
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(productActions, dispatch),
    userAction : bindActionCreators(userActions,dispatch)
  };
}
export default (connect(mapStateToProps, mapDispatchToProps))(AllProductsPage);
