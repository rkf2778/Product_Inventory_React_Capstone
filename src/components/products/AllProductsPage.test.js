import React from 'react'
import AllProductsPage from './AllProductsPage'
import ProductList from "./ProductList";
import Pagination from './Pagination'
import toJson from 'enzyme-to-json';
import {shallow,mount} from 'enzyme'
import { Provider } from "react-redux";
import configureStore from "../redux/store/configureStore";

describe('All Products Page Snapshot', () => {
    let mountwrapper;
    const store = configureStore();

    beforeEach(()=>{
        mountwrapper=mount(<Provider store={store}><AllProductsPage/></Provider>);
    })    

    afterEach(()=>{
        mountwrapper.unmount();
    })

    it('renders correctly', () => {
        expect(mountwrapper).toMatchSnapshot();
    });

});


describe('All elements renders correctly', () => {
    const store = configureStore();
    let wrapper = shallow(<Provider store={store}><AllProductsPage/></Provider>)

    it('shows the debug code',()=>{
        expect(wrapper.find(ProductList).length).toEqual(0)
    })

});