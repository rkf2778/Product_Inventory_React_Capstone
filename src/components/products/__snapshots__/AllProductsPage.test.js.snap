// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Products Page Snapshot renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(AllProductsPage)>
    <AllProductsPage
      actions={
        Object {
          "addProduct": [Function],
          "addView": [Function],
          "deleteProduct": [Function],
          "loadProduct": [Function],
          "updateProduct": [Function],
        }
      }
      products={Array []}
      userAction={
        Object {
          "addUser": [Function],
          "loadUser": [Function],
        }
      }
      users={Array []}
    >
      <div>
        <div
          className="container"
          style={
            Object {
              "display": "flex",
              "paddingTop": "30px",
            }
          }
        >
          <span
            style={
              Object {
                "marginLeft": "auto",
              }
            }
          >
            <input
              onChange={[Function]}
              type="text"
            />
             
             
             
            <FiSearch
              size="20px"
            >
              <IconBase
                attr={
                  Object {
                    "fill": "none",
                    "stroke": "currentColor",
                    "strokeLinecap": "round",
                    "strokeLinejoin": "round",
                    "strokeWidth": "2",
                    "viewBox": "0 0 24 24",
                  }
                }
                size="20px"
              >
                <svg
                  fill="none"
                  height="20px"
                  size="20px"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 24 24"
                  width="20px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="11"
                    cy="11"
                    key="0"
                    r="8"
                  />
                  <line
                    key="1"
                    x1="21"
                    x2="16.65"
                    y1="21"
                    y2="16.65"
                  />
                </svg>
              </IconBase>
            </FiSearch>
             
          </span>
        </div>
        <div
          className="container"
          style={
            Object {
              "alignItems": "space-between",
              "display": "flex",
              "justifyContent": "flex-end",
              "paddingTop": "6px",
            }
          }
        >
          <label
            style={
              Object {
                "color": "white",
                "padding": "0px 5px 0px 2px",
              }
            }
          >
            <input
              defaultChecked={true}
              onClick={[Function]}
              type="checkbox"
            />
             
            Quantity
          </label>
          <label
            style={
              Object {
                "color": "white",
                "padding": "0px 5px 0px 2px",
              }
            }
          >
            <input
              defaultChecked={true}
              onClick={[Function]}
              type="checkbox"
            />
             
            Price 
          </label>
          <label
            style={
              Object {
                "color": "white",
                "padding": "0px 5px 0px 2px",
              }
            }
          >
            <input
              defaultChecked={true}
              onClick={[Function]}
              type="checkbox"
            />
             
            Manufacturer 
          </label>
        </div>
        <hr />
        <div
          style={
            Object {
              "minHeight": "100vh",
            }
          }
        >
          <ProductList
            data={
              Object {
                "manufacturer": true,
                "price": true,
                "quantity": true,
              }
            }
            products={Array []}
            toggleManufacturer={[Function]}
            togglePrice={[Function]}
            toggleQuantity={[Function]}
          >
            <Container
              fluid={false}
            >
              <div
                className="container"
              />
            </Container>
          </ProductList>
          <br />
          <Container
            fluid={false}
          >
            <div
              className="container"
            >
              <Row
                noGutters={false}
              >
                <div
                  className="row"
                >
                  <Col>
                    <div
                      className="col"
                    />
                  </Col>
                  <Col
                    lg="auto"
                    md="auto"
                    sm="auto"
                    xs="auto"
                  >
                    <div
                      className="col-lg-auto col-md-auto col-sm-auto col-auto"
                    >
                      <Pagination
                        paginate={[Function]}
                        postsPerPage={9}
                        totalPosts={0}
                      >
                        <nav>
                          <ul
                            className="pagination"
                          />
                        </nav>
                      </Pagination>
                    </div>
                  </Col>
                  <Col>
                    <div
                      className="col"
                    />
                  </Col>
                </div>
              </Row>
            </div>
          </Container>
        </div>
      </div>
    </AllProductsPage>
  </Connect(AllProductsPage)>
</Provider>
`;
